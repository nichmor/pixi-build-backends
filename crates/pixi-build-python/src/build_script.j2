{% set PYTHON="%PYTHON%" if build_platform == "windows" else "$PYTHON" -%}
{% set SRC_DIR = manifest_root if editable else ("%SRC_DIR%" if build_platform == "windows" else "$SRC_DIR") -%}
{% set EDITABLE_OPTION = " --editable" if editable else "" -%}
{% set COMMON_OPTIONS = "-vv --no-deps --no-build-isolation" + EDITABLE_OPTION -%}

{{ PYTHON }} -m pip install --ignore-installed {{ COMMON_OPTIONS }} {{ SRC_DIR }}

{% if build_platform == "windows" -%}
if errorlevel 1 exit 1
{% endif %}
