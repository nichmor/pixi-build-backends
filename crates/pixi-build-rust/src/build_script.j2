{% if export_openssl -%}
export OPENSSL_DIR="$PREFIX"
{% endif %}

cargo --config net.git-fetch-with-cli=true install --locked --root $PREFIX --path {{ source_dir }} --no-track {{ extra_args | join(" ") }} --force
