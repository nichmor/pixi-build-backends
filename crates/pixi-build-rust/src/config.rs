use serde::Deserialize;

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct RustBackendConfig {
    /// Extra args to pass for cargo
    pub extra_args: Vec<String>,
}
