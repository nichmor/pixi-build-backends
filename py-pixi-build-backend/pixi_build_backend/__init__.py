"""
Python implementation of pixi-build-python backend using Python bindings.
"""

# __all__ = [
#     "PythonGenerator",
#     "PythonBackendConfig",
#     "BuildScriptContext",
#     "Installer",
#     "BuildPlatform",
#     "ProjectModelV1",
# ]
