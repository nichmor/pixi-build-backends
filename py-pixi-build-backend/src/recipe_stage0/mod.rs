pub mod conditional;
pub mod recipe;
pub mod requirements;
pub mod utils;
