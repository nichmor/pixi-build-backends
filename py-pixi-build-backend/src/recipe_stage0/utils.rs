use pyo3::{PyResult, pyfunction};
use pyproject_toml::PyProjectToml;
use rattler_conda_types::package::EntryPoint;
use std::collections::HashMap;

/// Parse entry points from a pyproject.toml content string
#[pyfunction]
pub fn parse_entry_points_from_pyproject(content: String) -> PyResult<Vec<String>> {
    let pyproject: PyProjectToml = toml::from_str(&content).map_err(|e| {
        pyo3::exceptions::PyValueError::new_err(format!("Invalid pyproject.toml: {}", e))
    })?;

    let entry_points = pyproject
        .project
        .as_ref()
        .and_then(|p| p.scripts.as_ref())
        .map(|scripts| {
            scripts
                .iter()
                .map(|(name, entry_point)| format!("{} = {}", name, entry_point))
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();

    Ok(entry_points)
}

/// Parse entry points from a dictionary of scripts
#[pyfunction]
pub fn parse_entry_points_from_scripts(scripts: HashMap<String, String>) -> PyResult<Vec<String>> {
    let entry_points: Result<Vec<EntryPoint>, _> = scripts
        .into_iter()
        .map(|(name, entry_point)| format!("{} = {}", name, entry_point).parse())
        .collect();

    match entry_points {
        Ok(entry_points) => Ok(entry_points.into_iter().map(|ep| ep.to_string()).collect()),
        Err(_) => Err(pyo3::exceptions::PyValueError::new_err(
            "Invalid entry point format",
        )),
    }
}

/// Validate an entry point string
#[pyfunction]
pub fn validate_entry_point(entry_point: String) -> PyResult<bool> {
    match entry_point.parse::<EntryPoint>() {
        Ok(_) => Ok(true),
        Err(_) => Ok(false),
    }
}

/// Parse a single entry point string and return its components
#[pyfunction]
pub fn parse_entry_point(entry_point: String) -> PyResult<(String, String, Option<String>)> {
    let ep: EntryPoint = entry_point.parse().map_err(|e| {
        pyo3::exceptions::PyValueError::new_err(format!("Invalid entry point: {}", e))
    })?;

    Ok((ep.command, ep.module, Some(ep.function)))
}

/// Create an entry point string from components
#[pyfunction]
pub fn create_entry_point(name: String, module: String, function: Option<String>) -> String {
    match function {
        Some(function) => format!("{} = {}:{}", name, module, function),
        None => format!("{} = {}", name, module),
    }
}
