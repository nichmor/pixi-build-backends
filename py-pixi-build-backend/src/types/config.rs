use std::path::{Path, PathBuf};

use pixi_build_backend::generated_recipe::BackendConfig;
use pyo3::{Py, PyAny, PyObject, Python, pyclass, pymethods, types::PyAnyMethods};
use pythonize::pythonize;
use serde::{Deserialize, de::DeserializeOwned};
// use serde::{de::DeserializeOwned, Deserialize, Deserializer};

#[derive(Deserialize, Clone)]
#[pyclass]
pub struct PyBackendConfig {
    debug_dir: Option<PathBuf>,
}

#[pymethods]
impl PyBackendConfig {
    #[new]
    fn new(debug_dir: Option<PathBuf>) -> Self {
        PyBackendConfig { debug_dir }
    }

    fn debug_dir(&self) -> Option<&Path> {
        BackendConfig::debug_dir(self)
    }
}

impl BackendConfig for PyBackendConfig {
    fn debug_dir(&self) -> Option<&Path> {
        self.debug_dir.as_deref()
    }
}
