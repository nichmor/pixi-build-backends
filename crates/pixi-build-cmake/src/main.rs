mod build_script;
mod config;

use std::path::Path;

use build_script::{BuildPlatform, BuildScriptContext};
use config::CMakeBackendConfig;
use miette::IntoDiagnostic;
use pixi_build_backend::{
    compilers::{Language, compiler_requirement, default_compiler},
    generated_recipe::{GenerateRecipe, GeneratedRecipe},
    intermediate_backend::IntermediateBackendInstantiator,
};
use rattler_conda_types::{MatchSpec, PackageName, ParseStrictness, Platform};
use recipe_stage0::{matchspec::PackageDependency, recipe::Script};

#[derive(Default, Clone)]
pub struct CMakeGenerator {}

impl GenerateRecipe for CMakeGenerator {
    type Config = CMakeBackendConfig;

    fn generate_recipe(
        &self,
        model: &pixi_build_types::ProjectModelV1,
        config: &Self::Config,
        manifest_root: std::path::PathBuf,
        host_platform: rattler_conda_types::Platform,
    ) -> miette::Result<GeneratedRecipe> {
        let mut generated_recipe =
            GeneratedRecipe::from_model(model.clone(), manifest_root.clone());

        // we need to add compilers

        let requirements = &mut generated_recipe.recipe.requirements;

        let resolved_requirements = requirements.resolve(Some(&host_platform));

        // Ensure the compiler function is added to the build requirements
        // only if a specific compiler is not already present.
        // TODO: Correctly, we should ask cmake to give us the language used in the project
        // instead of assuming C++.
        let language_compiler = default_compiler(&host_platform, &Language::Cxx.to_string());

        let build_platform = Platform::current();

        if !resolved_requirements
            .build
            .contains_key(&PackageName::new_unchecked(language_compiler))
        {
            if build_platform.is_windows() {
                // on windows, we use newer vs2019 compiler

                if !resolved_requirements
                    .build
                    .contains_key(&PackageName::new_unchecked("vs2017"))
                    && !resolved_requirements
                        .build
                        .contains_key(&PackageName::new_unchecked("vs2019"))
                {
                    let vss = MatchSpec::from_str("vs2019", ParseStrictness::Strict)
                        .expect("Failed to parse vs2019 match spec");
                    requirements
                        .build
                        .push(PackageDependency::Binary(vss).into());
                }
            } else {
                // otherwise we default to cmpi
                requirements
                    .build
                    .push(compiler_requirement(&Language::Cxx));
            }
        }

        // add necessary build tools
        for tool in ["cmake", "ninja"] {
            let tool_name = PackageName::new_unchecked(tool);
            if !resolved_requirements.build.contains_key(&tool_name) {
                requirements.build.push(tool.parse().into_diagnostic()?);
            }
        }

        // Check if the host platform has a host python dependency
        // This is used to determine if we need to the cmake argument for the python executable
        let has_host_python = resolved_requirements.contains(&PackageName::new_unchecked("python"));

        let build_script = BuildScriptContext {
            build_platform: if build_platform.is_windows() {
                BuildPlatform::Windows
            } else {
                BuildPlatform::Unix
            },
            source_dir: manifest_root.display().to_string(),
            extra_args: config.extra_args.clone(),
            has_host_python,
        }
        .render();

        generated_recipe.recipe.build.script = Script {
            content: build_script,
            env: config.env.clone(),
            ..Default::default()
        };

        Ok(generated_recipe)
    }

    fn build_input_globs(config: &Self::Config, _workdir: impl AsRef<Path>) -> Vec<String> {
        [
            // Source files
            "**/*.{c,cc,cxx,cpp,h,hpp,hxx}",
            // CMake files
            "**/*.{cmake,cmake.in}",
            "**/CMakeFiles.txt",
        ]
        .iter()
        .map(|s| s.to_string())
        .chain(config.extra_input_globs.clone())
        .collect()
    }
}

#[tokio::main]
pub async fn main() {
    if let Err(err) =
        pixi_build_backend::cli::main(IntermediateBackendInstantiator::<CMakeGenerator>::new).await
    {
        eprintln!("{err:?}");
        std::process::exit(1);
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::*;

    use indexmap::IndexMap;
    use pixi_build_types::ProjectModelV1;

    #[test]
    fn test_input_globs_includes_extra_globs() {
        let config = CMakeBackendConfig {
            extra_input_globs: vec!["custom/*.c".to_string()],
            ..Default::default()
        };

        let result = CMakeGenerator::build_input_globs(&config, PathBuf::new());

        insta::assert_debug_snapshot!(result);
    }

    #[macro_export]
    macro_rules! project_fixture {
        ($($json:tt)+) => {
            serde_json::from_value::<ProjectModelV1>(
                serde_json::json!($($json)+)
            ).expect("Failed to create TestProjectModel from JSON fixture.")
        };
    }

    #[test]
    fn test_cxx_is_in_build_requirements() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let generated_recipe = CMakeGenerator::default()
            .generate_recipe(
                &project_model,
                &CMakeBackendConfig::default(),
                PathBuf::from("."),
                Platform::Linux64,
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe, {
        ".source[0].path" => "[ ... path ... ]",
        ".build.script" => "[ ... script ... ]",
        });
    }

    #[test]
    fn test_env_vars_are_set() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let env = IndexMap::from([("foo".to_string(), "bar".to_string())]);

        let generated_recipe = CMakeGenerator::default()
            .generate_recipe(
                &project_model,
                &CMakeBackendConfig {
                    env: env.clone(),
                    ..Default::default()
                },
                PathBuf::from("."),
                Platform::Linux64,
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe.build.script,
        {
            ".content" => "[ ... script ... ]",
        });
    }

    #[test]
    fn test_has_python_is_set_in_build_script() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    },
                    "hostDependencies": {
                        "python": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let generated_recipe = CMakeGenerator::default()
            .generate_recipe(
                &project_model,
                &CMakeBackendConfig::default(),
                PathBuf::from("."),
                Platform::Linux64,
            )
            .expect("Failed to generate recipe");

        // we want to check that
        // -DPython_EXECUTABLE=$PYTHON is set in the build script
        insta::assert_yaml_snapshot!(generated_recipe.recipe.build,

            {
            ".script.content" => insta::dynamic_redaction(|value, _path| {
                dbg!(&value);
                // assert that the value looks like a uuid here
                assert!(value
                    .as_slice()
                    .unwrap()
                    .iter()
                    .any(|c| c.as_str().unwrap().contains("-DPython_EXECUTABLE"))
                );
                "[content]"
            })
        });
    }

    #[test]
    fn test_cxx_is_not_added_if_gcc_is_already_present() {
        let project_model = project_fixture!({
            "name": "foobar",
            "version": "0.1.0",
            "targets": {
                "defaultTarget": {
                    "runDependencies": {
                        "boltons": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    },
                    "buildDependencies": {
                        "gxx": {
                            "binary": {
                                "version": "*"
                            }
                        }
                    }
                },
            }
        });

        let generated_recipe = CMakeGenerator::default()
            .generate_recipe(
                &project_model,
                &CMakeBackendConfig::default(),
                PathBuf::from("."),
                Platform::Linux64,
            )
            .expect("Failed to generate recipe");

        insta::assert_yaml_snapshot!(generated_recipe.recipe, {
        ".source[0].path" => "[ ... path ... ]",
        ".build.script" => "[ ... script ... ]",
        });
    }
}
