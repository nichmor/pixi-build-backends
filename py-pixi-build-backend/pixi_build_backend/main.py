"""
Main entry point for Python backend implementation.
"""

import sys
import asyncio
from pixi_build_backend.types.generated_recipe import GenerateRecipeProtocol
from pixi_build_backend.pixi_build_backend import py_main, PyGenerateRecipe


async def main_entry_point(instance: GenerateRecipeProtocol):
    """Main entry point for the build backend"""
    py_generator = PyGenerateRecipe(instance)

    # Remove python name from argv
    args = sys.argv[1:] if len(sys.argv) > 1 else []
    print("sys argv", sys.argv)

    try:
        await py_main(py_generator, sys.argv)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
