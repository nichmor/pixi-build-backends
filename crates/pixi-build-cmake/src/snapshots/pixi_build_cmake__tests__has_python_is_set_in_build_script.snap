---
source: crates/pixi-build-cmake/src/main.rs
expression: generated_recipe.recipe.build
---
number: ~
script:
  content:
    - ninja --version
    - cmake --version
    - ""
    - ""
    - "if [ ! -f \"$SRC_DIR/../build/build.ninja\" ]; then"
    - "    cmake $CMAKE_ARGS \\"
    - "          -GNinja \\"
    - "          -DCMAKE_BUILD_TYPE=Release \\"
    - "          -DCMAKE_INSTALL_PREFIX=$PREFIX \\"
    - "          -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \\"
    - "          -DPython_EXECUTABLE=$PYTHON \\"
    - "          -DBUILD_SHARED_LIBS=ON \\"
    - "           \\"
    - "          -B $SRC_DIR/../build \\"
    - "          -S \".\""
    - fi
    - cmake --build $SRC_DIR/../build --target install
  env: {}
